documentation/specs/static-graph.md

```diff
@@ -143,49 +139,100 @@ This is a proposal for what the public API may look like:
+- Proj4 is built with targets $B, A, C, D$. On multiple parents, the incoming targets get concatenated. The order does not matter, as MSBuild has non-deterministic p2p ordering.
+- Proj3 and Proj2 build with $B, A$, as specified by the rule in Proj1.
+- Proj1 builds with A, because it's the root of the graph.
 
-For projects which depend on other projects, the MSBuild task will look up the target result in the cache. If the target is missing from the cache, an error will be logged and the build will fail. If the project is calling into itself either via `CallTarget` or via `MSBuild` with a different set of global properties, this will be always allowed (support for cross-targeting).
+```
+            A+-->Proj1   A->B, A
+                 /    \
+           B, A /      \ B, A
+               /        \
+              v          v
+  A->B, A   Proj3      Proj2   A->C, D
+             \            /
+        B, A  \          /  C, D
+               \        /
+                \      /
+                 v    v
+                 Proj4   A->B, A
+```
+
+The common project reference protocols (Build, Rebuild, Restore, Clean) will be specified by the common props and targets file in the msbuild repository. Other SDKs can implement their own protocols (e.g. ASPNET implementing Publish).
+
+Here are the rules for the common protocols:
+
+$Build \to GetTargetFrameworks, \langle default \rangle, GetNativeManifest, GetCopyToOutputDirectoryItems$
+
+The default target (represented here as $\langle default \rangle$) is resolved for each project.
+
+$Clean \to GetTargetFrameworks, Clean$
+
+$Rebuild \to \$Clean, \$Build$
+
+$Rebuild$ maps to two referenced protocols. $\$Clean$ and $\$Build$ get substituted with the contents of the $Clean$ and $Build$ protocols.
+
+Restore is a composition of two rules:
+- $Restore \to \_IsProjectRestoreSupported, \_GenerateRestoreProjectPathWalk, \_GenerateRestoreGraphProjectEntry$
 
-This makes the P2P contract explicit as it requires both projects to declare targets which are part of the contract.
+- $\_GenerateRestoreProjectPathWalk \to \_IsProjectRestoreSupported, \_GenerateRestoreProjectPathWalk, \_GenerateRestoreGraphProjectEntry$
+
+**Open Issue:** Restore is a bit complicated, and we may need new concepts to represent it. The root project calls the recursive $\_GenerateRestoreProjectPathWalk$ on itself to collect the children closure, and then after the recursion call returns (after having walked the graph), it calls the other targets on each returned child in a non-recursive manner. So the above protocol is not a truthful representation of what happens, but it correctly captures all targets called on each node in the graph.
+
+**Open Issue:** How do we differentiate between targets called on the outer build project and targets called on the inner build projects?
+
+We'll represent the project reference protocols as items in MSBuild:
```

Comment:
I'd go so far as to say that if we go down this road there _must_ be a property. Otherwise it'd be absurdly hard to extend behavior.

Reply:
Yes, we discussed that everything *we* defined will have a property. Custom SDK owners can do whatever they want, but our pattern will be to define a property in the common props, then define the items in the common targets so that anything we own can be extended.

Spec-wise though it's not really a requirement that a property be used. It's just a pattern we'll hold ourselves to for extensitiblity.